content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,209 @@ Like a segment tree on values rather than indices.
+
+### Approaching the Solution
+
+One way to solve the problem is to go to the leaf node
+and perform a binary search for the number of indices less than 𝑟
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+### A different way to find the Index of 𝑟 in the list of Vertices
+
+Instead of binary searching on the leaf, we update 𝑟 as we recurse down the
+tree.
+If we are able to determine the position (index) of 𝑟 in the left and right
+children of a node
+We can recurse down the tree and determine its position in the leaf node.
+
+### Determining the Position in Child Nodes
+
+To find the position of 𝑟 in a node's left and right children, we need to
+determine how many indices are smaller than the middle value (mid) and precede
+𝑟.
+This can be done using a prefix sum.
+
+Let's define:
+c[i] = as 1 if index[i] is smaller than mid otherwise 0
+prefixB[i] as prefix sum of c[i]
+
+Formally
+
+$$
+c[i] = index[i] < mid ? 1 : 0;
+prefixB[i] = prefixB[i - 1] + c[i]
+$$
+
+### Updating 𝑟 as We Recurse
+
+To know the value of 𝑟 if we recurse left, we use prefixB[r]
+If we recurse right, we use 𝑟 - prefixB[r]
+
+## Solving the second type of query (Given a range l, r find the k smallest element)
+
+We will determine whether the answer for a given node is in the left or the
+right segment.
+We can calculate how many times the elements within the segments' ranges appear
+in our range (l, r) using our first type of query.
+Note that this also works for non-leaf nodes using the following formula:
+
+$$
+occurrences(l, r) = r - l
+$$
+<Info title="Simular">
+This is simular to counting how many times a value appears up to index 𝑅 in our previous query. We did this by using the new 𝑅 value at the leaf node. But now, we consider the difference between the updated 𝑅 and 𝐿
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+left_occurrences = prefixB[r] - prefixB[l]
+$$
+
+<Info title="Left occurrences">
+  Note that left_occurrences is the amount of indices between l and r that their value is less than mid
+
+</Info>
+
+- If left_occurrences left_occurrences is greater or equal to k, it means the k-th smallest element is in
+  the left subtree. Therefore, we update our range and recurse into the left
+  child
+- If left_occurrences is less than k, it means the
+  k-th smallest element is in the right subtree. We adjust k by subtracting
+  left_occurrences from k, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update l, r accordingly when we go left or right
+</Info>
+
+the answer then will be the value of the node we end up on (leaf)
+
+## implemention
+
+<CPPSection>
+
+```cpp
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefixB;
+
+	Segment(int l, int r, vector<pair<int, int>> indices)
```

Comment:
does the vector have to be pass by value

Reply:
No, it does not have to be passed by value. Switching to pass-by-reference will not change the time complexity, but it will avoid unnecessary memory allocation.

Reply:
I have reviewed the code again. A copy of the vector must be stored inside the Segment. Therefore, if we switch to passing by reference, we will still need to make a copy of the vector in the line indices(indices). Using pass-by-value and std::move achieves the same result.